### ISSUE TITLE
Archives are not reproducible across Windows/Linux

### ISSUE THREAD

**Issue by ap-actoron** [gradle/gradle#8051]
When using ```reproducibleFileOrder=true``` in, e.g., a Jar task, building the same project on Windows and on Linux creates jars with different file orders due to Windows being case-insensitive.

### Expected Behavior
Jar entries should be in the same order when built on Windows or Linux when  ```reproducibleFileOrder=true```.

### Current Behavior
File order differs on Windows and Linux, e.g.:
```
built-on-windows.jar
A1.txt
a2.txt
B1.txt

built-on-linux.jar
A1.txt
B1.txt
a2.txt
```

### Cause of Bug
The implementation of org.gradle.api.internal.file.collections.ReproducibleDirectoryWalker relies on ```java.io.File.compareTo()```, which is OS-dependent as also stated in its API: *"Compares two abstract pathnames lexicographically. The ordering defined by this method depends upon the underlying system. On UNIX systems, alphabetic case is significant in comparing pathnames; on Microsoft Windows systems it is not."*

### Suggested Fix
Force case (in-)sensitive file ordering on all platforms. Therefore replace
```
File[] children = file.listFiles();
if (children != null) {
	Arrays.sort(children);
}
```

with something like:
```
File[] children = file.listFiles();
if (children != null) {
	Arrays.sort(children, new Comparator<File>() {
		@Override
		public int compare(File o1, File o2) {
			//return o1.getName().toLowerCase().compareTo(o2.getName().toLowerCase()); // case insensitive on all platforms
			return o1.getName().compareTo(o2.getName()); // case sensitive on all platforms
    		}
	});
}

**Comment by big-guy** [comment]
@ap-actoron would you be willing to submit a PR for this change?

**Comment by ap-actoron** [comment]
@big-guy Sure. As I've never done a PR for gradle before: anything I need to adhere to in terms of coding style, commit message etc.?

**Comment by big-guy** [comment]
Yep, [some of that is here](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).  For the most part, it's follow what you find and accept the defaults of the IDE.  We have some Checkstyle/Codenarc rules that catch other things.

For this issue, I think you're in the right area WRT `ReproducibleDirectoryWalker`.  We can get a case-insensitive file comparator from commons-io, I think.  I would expect some changes to `ReproducibleArchivesIntegrationTest` to demonstrate a jar with different cased entries (files, directories and sub-directories) and how that interacts with the filesystem case-sensitivity.  You can look at test case `reproducible #taskName for directory - #files` for inspiration.

You can run these tests with `./gradlew core:integTest --tests "ReproducibleArchivesIntegrationTest"` or from the IDE (we all use IntelliJ).  Please be sure to run `./gradlew <subproject>:check` on the subprojects that you change when you submit the PR.  We'll look over it and then trigger a full test build.

Let me know if you have any questions. Thanks in advance.

**Comment by miokowpak** [comment]
@ap-actoron Are you still planning on submitting a PR for this?  I ran into the same issue today, and will be happy to take a stab at making the change if you have no plans to do this anytime soon.

**Comment by ap-actoron** [comment]
@miokowpak I'm not currently working on a PR and would be happy if you do it.

**Comment by wolfs** [comment]
Fixed by #9416.
